### PR TITLE
feat: dash event msg box handling

### PR DESCRIPTION
**Pull request by lfordyce**

### CODE REVIEW

File: packager/media/formats/mp4/mp4_media_parser.h [hunk 1]
```diff
@@ -36,6 +36,7 @@ class MP4MediaParser : public MediaParser {
   void Init(const InitCB& init_cb,
             const NewMediaSampleCB& new_media_sample_cb,
             const NewTextSampleCB& new_text_sample_cb,
+            const DASHEventMessageBoxCB& event_message_cb,
```
Comment (chenda6) [inline packager/media/formats/mp4/mp4_media_parser.h:39]: The `emsg` boxes only occur in mp4 containers, so we can add the callback to this class only without having to change the `MediaParser::Init()` interface.  This allows the whole class hierarchy and other files remain untouched.

For instance, one way we can accomplish this is by adding a member function and then calling it in the demuxer (for the appropriate container type).
```
void MP4MediaParser::SetEventMessageBoxCB(const DASHEventMessageBoxCB& event_message_cb) {
  event_message_cb_ = event_message_cb;
}
```

File: packager/live_packager_test.cc [hunk 2]
```diff
@@ -1155,17 +1176,26 @@ TEST_P(LivePackagerTestReEncrypt, VerifyReEncryption) {
...
   CHECK(parser_noenc_->Parse(expected_buf.data(), expected_buf.size()));
   auto& expected_samples = parser_noenc_->GetSamples();
+  auto& expected_emsg_samples = parser_noenc_->GetEmsgSamples();
 
   CHECK(parser_enc_->Parse(actual_buf.Data(), actual_buf.Size()));
   auto& actual_samples = parser_enc_->GetSamples();
+  auto& actual_emsg_samples = parser_enc_->GetEmsgSamples();
 
   CHECK_EQ(expected_samples.size(), actual_samples.size());
+  CHECK_EQ(expected_emsg_samples.size(), actual_emsg_samples.size());
```
Comment (chenda6) [inline packager/live_packager_test.cc:1186]: Could you please add checks here for:

```
expected_samples.size() > 0
```
and 
```
expected_emsg_samples.size() > 0
```

**Comment by chenda6** [comment]
Really good work figuring out how to preserve the `emsg` boxes!

File: packager/live_packager_test.cc [hunk 3]
```diff
@@ -1155,17 +1176,26 @@ TEST_P(LivePackagerTestReEncrypt, VerifyReEncryption) {
...
       expected_samples.begin(), expected_samples.end(), actual_samples.begin(),
       actual_samples.end(), [](const auto& s1, const auto& s2) {
         return s1->data_size() == s2->data_size() &&
                0 == memcmp(s1->data(), s2->data(), s1->data_size());
       }));
+
+  CHECK(std::equal(expected_emsg_samples.begin(), expected_emsg_samples.end(),
+                   actual_emsg_samples.begin(), actual_emsg_samples.end(),
+                   [](const auto& s1, const auto& s2) {
+                     return s1->GetID() == s2->GetID();
```
Comment (chenda6) [inline packager/live_packager_test.cc:1197]: Can we compare the entirety of `s1` and `s2`?

File: packager/app/muxer_factory.cc [hunk 4]
```diff
@@ -67,7 +69,11 @@ std::shared_ptr<Muxer> MuxerFactory::CreateMuxer(
         muxer = std::make_shared<mp4::MP4InitMuxer>(options);
         break;
       }
-      muxer = std::make_shared<mp4::MP4Muxer>(options);
+      {
+        auto mp4_muxer = std::make_shared<mp4::MP4Muxer>(options);
+        mp4_muxer->SetDashEventMessageHandler(dash_handler);
+        muxer = mp4_muxer;
```
Comment (chenda6) [inline packager/app/muxer_factory.cc:75]: `muxer = std::move(mp4_muxer);`
or even better:
`static_cast<mp4::MP4Muxer*>(muxer.get())->SetDashEventMessageHandler(dash_handler);`

**Comment by Vadym-Kupriyanchuk** [comment]
Overall looks good.
Items to consider:
-  Maybe we can use typedef/using for such types - std::shared_ptr<mp4::DashEventMessageHandler>, e.g. EmsgHandlePtr, this will make code more readable, especially with the shaka crazy rule for line width 
- the same for DashEventMessage

File: packager/live_packager_test.cc [hunk 5]
```diff
@@ -1155,33 +1179,57 @@ TEST_P(LivePackagerTestReEncrypt, VerifyReEncryption) {
...
                0 == memcmp(s1->data(), s2->data(), s1->data_size());
       }));
+
+  if (GetParam().emsg_processing) {
+    ASSERT_GT(expected_emsg_samples.size(), 0);
+    CHECK_EQ(expected_emsg_samples.size(), actual_emsg_samples.size());
+    CHECK(std::equal(expected_emsg_samples.begin(), expected_emsg_samples.end(),
+                     actual_emsg_samples.begin(), actual_emsg_samples.end(),
+                     [](const auto& s1, const auto& s2) {
+                       return (*s1.get()) == (*s2.get());
```
Comment (chenda6) [inline packager/live_packager_test.cc:1203]: FYI: This can be done as `return *s1 == *s2`;

**Review by chenda6 — APPROVED** [review]

**Pull request merged**